src/Monolog/Handler/StreamHandler.php

```diff
@@ -188,15 +191,27 @@ private function createDir(string $url): void
         }
 
         $dir = $this->getDirFromStream($url);
-        if (null !== $dir && !is_dir($dir)) {
-            $this->errorMessage = null;
-            set_error_handler([$this, 'customErrorHandler']);
-            $status = mkdir($dir, 0777, true);
-            restore_error_handler();
-            if (false === $status && !is_dir($dir)) {
-                throw new \UnexpectedValueException(sprintf('There is no existing directory at "%s" and it could not be created: '.$this->errorMessage, $dir));
+
+        $this->dirLock = tempnam(sys_get_temp_dir(), md5($dir . time()));
+
+        $resourceLock = fopen($this->dirLock, "r+");
+
+        //  Gets the exclusive lock for creating directory in parallel mode (eg: paratest)
+        if (true === flock($resourceLock, LOCK_EX)) {
```

Comment:
I don't quite understand what this solves. The way I see it it is going to create a different dirLock file in every process (as tempnam creates a unique filename) so there is no real protection between processes. I also don't see why this is stored as a property vs a simple local variable here.

One could perhaps acquire a lock on `dirname($dir)` before creating it, that might help, but if you have tests running in parallel which create/destroy the dir repeatedly you will anyway run into issues, the only fix for this is to run every test in its own directory.

Reply:
Yep, you are right. The resource lock must be the same for all concurrent processes, and should depend on only directory name. Directory lock variable can be local here. 

`but if you have tests running in parallel which create/destroy the dir repeatedly you will anyway run into issues`
yes, you are right, but it's not our case here.

we start paratest in 4-6 processes, it runs phpunit, phpunit uses monolog.
and no idea if phpunut drop monolog, which can be the problem of permanent create/destroy of test working directory..

Reply:
Could you see then?

Reply:
I don't quite understand what problem you are trying to fix then. The issue in https://github.com/Seldaek/monolog/issues/1678 is a different one IMO as it appears the dir gets wiped again.. 

Can you share what issue you are seeing? What error message? I'd really like to understand this better before adding layers and layers of locking that may not fix anything.